File: examples/clj/test/com/rpl/agent/basic/document_store_agent_test.clj
(ns com.rpl.agent.basic.document-store-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.document-store-agent :refer [DocumentStoreModule]]))

(deftest document-store-agent-test
  (System/gc)
  (testing "DocumentStoreAgent with simplified user profiles"
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc DocumentStoreModule {:tasks 1 :threads 1})

      (let [manager (aor/agent-manager ipc
                                       (rama/get-module-name
                                        DocumentStoreModule))
            agent   (aor/agent-client manager "DocumentStoreAgent")]

        (testing "creates and stores user profile"
          (let [result (aor/agent-invoke
                        agent
                        {:user-id         "test-user"
                         :profile-updates {:name        "Test User"
                                           :age         25
                                           :preferences {:theme "dark"}}})]
            (is (= "test-user" (:user-id result)))
            (is (= "Test User" (:name result)))
            (is (= 25 (:age result)))
            (is (= {:theme "dark"} (:preferences result)))))

        (testing "updates individual fields independently"
          (let [result (aor/agent-invoke
                        agent
                        {:user-id         "test-user"
                         :profile-updates {:age         30
                                           :preferences {:notifications true}}})]
            (is (= 30 (:age result)))
            (is (= "Test User" (:name result))) ; Name unchanged
            ;; Preferences should be merged
            (is (= {:theme "dark" :notifications true} (:preferences result)))))

        (testing "handles multiple users"
          (let [result (aor/agent-invoke
                        agent
                        {:user-id         "user2"
                         :profile-updates {:name        "Another User"
                                           :age         35
                                           :preferences {:theme "light"}}})]
            (is (= "user2" (:user-id result)))
            (is (= "Another User" (:name result)))
            (is (= 35 (:age result)))
            (is (= {:theme "light"} (:preferences result)))))))))

File: examples/java/src/test/java/com/rpl/agent/basic/DocumentStoreAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.HashMap;
import java.util.Map;
import org.junit.Test;

/**
 * Test class for DocumentStoreAgent demonstrating structured multi-field data storage.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>declareDocumentStore: Creating document storage with multiple fields
 *   <li>getStore: Accessing document stores from agent nodes
 *   <li>Store operations: getDocumentField, putDocumentField, updateDocumentField,
 *       containsDocumentField
 *   <li>HashMap usage for request and response data structures
 * </ul>
 */
public class DocumentStoreAgentTest {

  @Test
  public void testDocumentStoreAgent() throws Exception {
    // Tests document store operations with structured multi-field data using HashMap
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the agent module
      DocumentStoreAgent.DocumentStoreModule module = new DocumentStoreAgent.DocumentStoreModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("DocumentStoreAgent");

      // Test creating a user profile
      Map<String, Object> request = new HashMap<>();
      request.put("userId", "test-user");
      Map<String, Object> updates = new HashMap<>();
      updates.put("name", "Test User");
      updates.put("age", 30L);
      Map<String, Object> prefs = new HashMap<>();
      prefs.put("theme", "dark");
      prefs.put("newsletter", true);
      updates.put("preferences", prefs);
      request.put("profileUpdates", updates);

      @SuppressWarnings("unchecked")
      Map<String, Object> result = (Map<String, Object>) agent.invoke(request);

      assertNotNull("Result should not be null", result);
      assertEquals("User ID should match", "test-user", result.get("userId"));
      assertEquals("Name should match", "Test User", result.get("name"));
      assertEquals("Age should match", 30L, result.get("age"));

      @SuppressWarnings("unchecked")
      Map<String, Object> resultPrefs = (Map<String, Object>) result.get("preferences");
      assertNotNull("Preferences should not be null", resultPrefs);
      assertEquals("Theme should match", "dark", resultPrefs.get("theme"));
      assertEquals("Newsletter should match", true, resultPrefs.get("newsletter"));
    }
  }
}
